test/unittest/blas3/blas3_trsm_test.cpp

```diff
@@ -0,0 +1,95 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename blas3_gemm_test.cpp
+ *
+ **************************************************************************/
+
+#include "blas_test.hpp"
+
+template <typename scalar_t>
+using combination_t = std::tuple<int, int, char, char, char, char>;
+
+template <typename scalar_t>
+void run_test(const combination_t<scalar_t> combi) {
+  int m;
+  int n;
+  char transA;
+  char side;
+  char diag;
+  char triangle;
+  std::tie(m, n, transA, side, diag, triangle) = combi;
+
+  using data_t = utils::data_storage_t<scalar_t>;
+
+  scalar_t alpha = scalar_t{1};
+  const int lda = side == 'l' ? m : n;
+  const int ldb = m;
+  const int k = side == 'l' ? m : n;
+
+  const int sizeA = k * lda;
+  const int sizeB = n * ldb;
+
+  std::vector<scalar_t> A(sizeA);
+  std::vector<scalar_t> B(sizeB);
+
+  const scalar_t diagValue =
+      diag == 'u' ? scalar_t{1} : random_scalar(scalar_t{1}, scalar_t{10});
+
+  fill_triangle(A, k, k, lda, triangle, diagValue);
+  fill_random(B);
+
+  auto q = make_queue();
+  test_executor_t ex(q);
+  auto a_gpu = blas::make_sycl_iterator_buffer<scalar_t>(A, A.size());
+  auto b_gpu = blas::make_sycl_iterator_buffer<scalar_t>(B, B.size());
+
+  _trsm(ex, side, triangle, transA, diag, m, n, alpha, a_gpu, lda, b_gpu, ldb);
+
+  std::vector<data_t> X(B.size(), data_t{0});
+
+  auto x_gpu = blas::make_sycl_iterator_buffer<scalar_t>(X.size());
+  if (side == 'l') {
+    // Verification of the result. AX = B
+    _gemm(ex, transA, 'n', m, n, m, data_t{1}, a_gpu, lda, b_gpu, ldb,
+          data_t{0}, x_gpu, lda);
+  } else {
+    // Verification of the result. XA = B
+    _gemm(ex, 'n', transA, m, n, n, data_t{1}, b_gpu, ldb, a_gpu, lda,
+          data_t{0}, x_gpu, ldb);
+  }
+  // Copy the verification result to X
+  ex.get_policy_handler().wait(
+      ex.get_policy_handler().copy_to_host(x_gpu, X.data(), X.size()));
+
+  ASSERT_TRUE(utils::compare_vectors(X, B));
+  ex.get_policy_handler().wait();
+}
+
+const auto combi = ::testing::Combine(::testing::Values(4, 7, 12, 16),  // m
```

Comment:
Can you try more combinations with a larger size that can populate the entire GPU. See the sizes used in BLAS and try to add them also you need to consider LDA, etc

Reply:
I can, but the INTEL_GPU backend is not building for me so I am investigating that. If I have `M>16` gemm fails with incorrect workgroup size

Reply:
That might be some bug in the breaking and passing size to gemm. Gemm has a static_assert to avoid incorrect workgroup/tile sizes

Reply:
discussed internally. We need to create a bug report for GEMM when double type is used to match the cache line size with tile sizes

Reply:
I believe this can be resolved for now?

Reply:
The proper testing for size has not been added yet. The main type that we care about is float. Adding double for SYCL-BLAS is experimental. And we dont have any cases that uses double. That is the reason Double test path is disable by default. The test size added here is not adequate for float, so it requires more test to see the algorithm actually works. Once different sizes similar to GEMM has been added we can solve the case

Reply:
This is outdated. In the latest version we are testing 7, 16, 70, 300. It is easy to add more but I don't it is necessary for the tests. Maybe for benchmarks but not tests.

Reply:
For the test in order to caver various sizes can you add one small size the original one that you have one medium size the one you have added now, and one large size. This will make sure that the Work-item/Workgroup/ComputeUnit over/under population has been correctly considered in creating the global/local thread size

Reply:
* Which size should I add that is smaller than 7x7?
* Which size should I add that is larger than 300x300?

Reply:
let's say 7X7, 513x513, 1027x1027

Reply:
Test sizes updated